codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java

```diff
@@ -39,22 +39,30 @@
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
+import java.util.Comparator;
 import java.util.HashMap;
 import java.util.HashSet;
-import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.ConcurrentSkipListSet;
 
 /**
  * Simple implementation of {@link Http2Connection}.
  */
 public class DefaultHttp2Connection implements Http2Connection {
 
+    private static final Comparator<Http2Stream> STREAMID_COMPARATOR = new Comparator<Http2Stream>() {
+        @Override
+        public int compare(Http2Stream o1, Http2Stream o2) {
+            return o1.id() - o2.id();
+        }
+    };
+
     private final Set<Listener> listeners = new HashSet<Listener>(4);
     private final IntObjectMap<Http2Stream> streamMap = new IntObjectHashMap<Http2Stream>();
     private final ConnectionStream connectionStream = new ConnectionStream();
-    private final Set<Http2Stream> activeStreams = new LinkedHashSet<Http2Stream>();
+    private final Set<Http2Stream> activeStreams = new ConcurrentSkipListSet<Http2Stream>(STREAMID_COMPARATOR);
```

Comment:
@buchgr - I guess it depends on the access pattern.  I think it is worth investigating the following:

1. For low frequency methods (i.e [channelInactive](https://github.com/netty/netty/blob/master/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java#L199)) or methods that definitely close the stream (assuming these are also low frequency) this may be OK.

2. For methods that *may* close streams because of exceptions we could take the approach used in [DefaultHttp2LocalFlowController.java](https://github.com/netty/netty/blob/master/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java#L92).  The goal being to delay notification and modification until after the iteration.